opencog/atoms/core/Variables.h

```diff
@@ -132,6 +132,11 @@ struct Variables : public FreeVariables
 	VariableDeepTypeMap _deep_typemap;
 	VariableDeepTypeMap _fuzzy_typemap;
 
+	/// Return true iff all variables are well typed. For now only
+	/// simple types are supported, specifically if some variable is
+	/// simple typed NOTYPE, then it returns false.
```

Comment:
Well-typed, IMO, means the associated set of "well-formed" expressions of that type is non-empty. For instance the intersection between ConceptNodes and PredicateNodes is the empty set, because neither of them inherits from the other, thus it is ill-typed. I don't think we have atomese representation yet for type intersection so I can't write you an atomese example of that. However we do have C++ functions that calculates type intersection, and as you can see one of them deals with "shallow" atom types, the C++ `Type` class, and this one would return NOTYPE when just empty intersection is detected. See https://github.com/opencog/atomspace/blob/master/opencog/atomutils/TypeUtils.cc#L318